corporal/httpgateway/interceptor_ui.go

```diff
@@ -0,0 +1,136 @@
+package httpgateway
+
+import (
+	"bytes"
+	"devture-matrix-corporal/corporal/httphelp"
+	"devture-matrix-corporal/corporal/matrix"
+	"devture-matrix-corporal/corporal/policy"
+	"devture-matrix-corporal/corporal/userauth"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/sirupsen/logrus"
+)
+
+type UiAuthInterceptor struct {
+	policyStore                       *policy.Store
+	homeserverDomainName              string
+	userAuthChecker                   *userauth.Checker
+	sharedSecretAuthPasswordGenerator *matrix.SharedSecretAuthPasswordGenerator
+}
+
+func NewUiAuthInterceptor(
+	policyStore *policy.Store,
+	homeserverDomainName string,
+	userAuthChecker *userauth.Checker,
+	sharedSecretAuthPasswordGenerator *matrix.SharedSecretAuthPasswordGenerator,
+) *UiAuthInterceptor {
+	return &UiAuthInterceptor{
+		policyStore:                       policyStore,
+		homeserverDomainName:              homeserverDomainName,
+		userAuthChecker:                   userAuthChecker,
+		sharedSecretAuthPasswordGenerator: sharedSecretAuthPasswordGenerator,
+	}
+}
+
+func (me *UiAuthInterceptor) Intercept(r *http.Request) InterceptorResponse {
+	loggingContextFields := logrus.Fields{}
+
+	var payload map[string]interface{}
+
+	err := httphelp.GetJsonFromRequestBody(r, &payload)
+	if err != nil {
+		loggingContextFields["err"] = err.Error()
+		return createInterceptorErrorResponse(loggingContextFields, matrix.ErrorBadJson, "Bad input")
+	}
+
+	// This all has to be done manually since UI auth can have custom data anywhere
+	if payload["auth"] != nil {
```

Comment:
I think we need more checks for whether a key is in the map.

Example:

```go
auth, exists := payload["auth"]
if !exists {
	// Not an interactive auth request, but a request of some other kind.
	// Let's just make it go through as-is.
	return InterceptorResponse{
		Result:               InterceptorResultProxy,
		LoggingContextFields: loggingContextFields,
	}
}
```

Same thing for other things below:
- `auth["type"]`
- `auth["password"]`
- etc.

You already have some checks that confirm data is of the proper type (`pass, pass_ok := auth["password"].(string)`), which is good.

-------------

Another thing which would improve code readability (and you can see in my example in this comment) a lot is to avoid code-nesting, in favor of early returns: https://softwareengineering.stackexchange.com/a/18473

This can be applied at least to:

- `if payload["auth"] != nil {` (as seen on the example code in this comment)
- `if auth["type"].(string) == "m.login.password" {` (well, when it's `!=`)

.. and then we'll save ourselves 2 indentation levels.